subsys/shell/shell_mqtt.c

```diff
@@ -0,0 +1,775 @@
+
+	password.utf8 = (uint8_t *)MQTT_PASSWORD;
+	password.size = strlen(MQTT_PASSWORD);
+	username.utf8 = (uint8_t *)MQTT_USERNAME;
+	username.size = strlen(MQTT_USERNAME);
+
+	mqtt_client_init(&ctx->client_ctx);
+
+	/* MQTT client configuration */
+	ctx->client_ctx.broker = &ctx->broker;
+	ctx->client_ctx.evt_cb = mqtt_evt_handler;
+	ctx->client_ctx.client_id.utf8 = (uint8_t *)sh_mqtt->device_id;
+	ctx->client_ctx.client_id.size = strlen(sh_mqtt->device_id);
+	ctx->client_ctx.password = &password;
+	ctx->client_ctx.user_name = &username;
+	ctx->client_ctx.protocol_version = MQTT_VERSION_3_1_1;
+
+	/* MQTT buffers configuration */
+	ctx->client_ctx.rx_buf = ctx->buf.rx;
+	ctx->client_ctx.rx_buf_size = sizeof(ctx->buf.rx);
+	ctx->client_ctx.tx_buf = ctx->buf.tx;
+	ctx->client_ctx.tx_buf_size = sizeof(ctx->buf.tx);
+
+	/* MQTT transport configuration */
+#ifdef CONFIG_MQTT_LIB_WEBSOCKET
+	ctx->client_ctx.transport.type = MQTT_TRANSPORT_NON_SECURE_WEBSOCKET;
+#else
+	ctx->client_ctx.transport.type = MQTT_TRANSPORT_NON_SECURE;
+#endif
+
+#ifdef CONFIG_MQTT_LIB_WEBSOCKET
+	ctx->client_ctx.transport.websocket.config.host = CONFIG_SHELL_MQTT_SERVER_ADDR;
+	ctx->client_ctx.transport.websocket.config.url = "/mqtt";
+	ctx->client_ctx.transport.websocket.config.tmp_buf = ctx->temp_ws_rx_buf;
+	ctx->client_ctx.transport.websocket.config.tmp_buf_len = MQTT_LIB_WEBSOCKET_RECV_BUF_LEN;
+	ctx->client_ctx.transport.websocket.timeout = 5 * MSEC_PER_SEC;
+#endif
+
+#ifdef CONFIG_SHELL_MQTT_SOCK5_PROXY_ADDR
+	uint32_t sockaddr_size;
+
+	if (ctx->socks5_proxy.sa_family == AF_INET) {
+		sockaddr_size = sizeof(struct sockaddr_in);
+	} else {
+		sockaddr_size = sizeof(struct sockaddr_in6);
+	}
+
+	mqtt_client_set_proxy(ctx->client_ctx., &ctx->socks5_proxy, sockaddr_size);
+#endif
+}
+
+static void shell_mqtt_handle_error(struct shell_mqtt *ctx)
+{
+	if (ctx->state == SHELL_MQTT_OP_ERROR) {
+		(void)k_work_cancel_delayable_sync(&ctx->mqtt_work, &ctx->work_sync);
+		(void)k_work_cancel_delayable_sync(&ctx->send_work, &ctx->work_sync);
+		(void)mqtt_abort(&ctx->client_ctx);
+		clear_fds(ctx);
+		LOG_WRN("Retrying MQTT connect work after %d ms if network connected",
+			APP_SLEEP_MSECS);
+		ctx->state = SHELL_MQTT_OP_WAIT_NET;
+		net_conn_mgr_resend_status();
+	}
+}
+
+/* Prototype */
+static void shell_mqtt_connect_work(struct k_work *work);
+
+/*
+ * Work routine to process MQTT packet and keep alive MQTT connection
+ */
+static void shell_mqtt_process_work(struct k_work *work)
+{
+	int64_t remaining = APP_SLEEP_MSECS;
+	int64_t start_time = k_uptime_get();
+	struct shell_mqtt *ctx;
+	int rc;
+
+	ctx = CONTAINER_OF(work, struct shell_mqtt, mqtt_work);
+
+	/* Listen to the port for 500 ms as defined by APP_SLEEP_MSECS */
+	while (remaining > 0) {
+		if (wait(remaining)) {
+			rc = mqtt_input(&ctx->client_ctx);
+			if (rc != 0) {
+				LOG_ERR("mqtt_input error: %d", rc);
+				goto process_err;
+			}
+		}
+
+		rc = mqtt_live(&ctx->client_ctx);
```

Comment:
Reminding myself that its normal for `mqtt_live` to return `-EAGAIN`, when it isn't time to send the ping